lib/srv/db/cassandra_test.go

```diff
@@ -0,0 +1,312 @@
+			allowDbNames: []string{types.Wildcard},
+			allowDbUsers: []string{types.Wildcard},
+			dbName:       "cassandra",
+			dbUser:       "cassandra",
+		},
+		{
+			desc:         "has access to nothing",
+			user:         "alice",
+			role:         "admin",
+			allowDbNames: []string{},
+			allowDbUsers: []string{},
+			dbName:       "cassandra",
+			dbUser:       "cassandra",
+			err:          "access to db denied",
+		},
+		{
+			desc:         "no access to users",
+			user:         "alice",
+			role:         "admin",
+			allowDbNames: []string{types.Wildcard},
+			allowDbUsers: []string{},
+			dbName:       "cassandra",
+			dbUser:       "cassandra",
+			err:          "access to db denied",
+		},
+		{
+			desc:         "access allowed to specific user/database",
+			user:         "alice",
+			role:         "admin",
+			allowDbNames: []string{"metrics"},
+			allowDbUsers: []string{"cassandra"},
+			dbName:       "metrics",
+			dbUser:       "cassandra",
+		},
+		{
+			desc:         "access denied to specific user/database",
+			user:         "alice",
+			role:         "admin",
+			allowDbNames: []string{"metrics"},
+			allowDbUsers: []string{"alice"},
+			dbName:       "cassandra",
+			dbUser:       "cassandra",
+			err:          "access to db denied",
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.desc, func(t *testing.T) {
+			// Create user/role with the requested permissions.
+			testCtx.createUserAndRole(ctx, t, test.user, test.role, test.allowDbUsers, test.allowDbNames)
+
+			// Try to connect to the database as this user.
+			dbConn, err := testCtx.cassandraClient(ctx, test.user, "cassandra", test.dbUser)
+			if test.err != "" {
+				require.Error(t, err)
+				require.Contains(t, err.Error(), test.err)
+				return
+			}
+			require.NoError(t, err)
+
+			// Execute a query.
+			var clusterName string
+			err = dbConn.Query("select cluster_name from system.local").Scan(&clusterName)
+			require.NoError(t, err)
+			require.Equal(t, "Test Cluster", clusterName)
+
+			// Disconnect.
+			dbConn.Close()
+		})
+	}
+}
+
+func TestAccessCassandraHandshake(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		protocolVersion primitive.ProtocolVersion
+	}{
+		{
+			protocolVersion: primitive.ProtocolVersion5,
+		},
+		{
+			protocolVersion: primitive.ProtocolVersion4,
+		},
+		{
+			protocolVersion: primitive.ProtocolVersion3,
+		},
+	}
+	ctx := context.Background()
+	testCtx := setupTestContext(ctx, t, withCassandra("cassandra"))
+	go testCtx.startHandlingConnections()
+	teleportUser := "alice"
+
+	for _, tc := range tests {
+		t.Run(tc.protocolVersion.String(), func(t *testing.T) {
+			t.Run("successful", func(t *testing.T) {
+				testCtx.createUserAndRole(ctx, t, teleportUser, "admin", []string{"cassandra"}, []string{types.Wildcard})
+				cqlRawClient, err := testCtx.cassandraRawClient(ctx, teleportUser, "cassandra", "cassandra")
+				require.NoError(t, err)
+				err = cqlRawClient.InitiateHandshake(tc.protocolVersion, 101)
```

Comment:
Can you replace the "magic numbers" with constants? That should make it easier to understand what all those numbers mean.